api/services/content/content.proto

```diff
@@ -0,0 +1,156 @@
+}
+
+message InfoRequest {
+	string digest = 1 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+}
+
+message InfoResponse {
+	// Digest is the hash identity of the blob.
+	string digest = 1 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+
+	// Size is the total number of bytes in the blob.
+	int64 size = 2;
+
+	// CommittedAt provides the type at which the blob was committed.
+	google.protobuf.Timestamp committed_at = 3;
+}
+
+// ReadRequest defines the 
+message ReadRequest {
+	// Digest is the hash identity to read.
+	string digest = 1 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+
+	// Offset specifies the number of bytes in the blob to begin the read. If
+	// zero or less, the read will be from the start.
+	int64 offset = 2;
+
+	// size is the total size of the read. If zero, the entire blob will be
+	// returned by the service.
+	int64 size = 3;
+}
+
+// ReadResponse carries byte data for a read request.
+message ReadResponse {
+	int64 offset = 1; // offset of the returned data
+	bytes data = 2; // actual data
+}
+
+
+// WriteAction sets the action for a WriteRequst.
+enum WriteAction {
+	option (gogoproto.goproto_enum_prefix) = false;
+	option (gogoproto.enum_customname) = "WriteAction";
+	
+	// WriteActionWrite sets the action for the write request to write data.
+	//
+	// Any data included will be written at the provided offset. The
+	// transaction will be left open for further writes.
+	//
+	// This is the default.
+	WRITE = 0 [(gogoproto.enumvalue_customname)="WriteActionWrite"];
+
+	// WriteActionCommit will write any outstanding data in the message and
+	// commit the write, storing it under the digest.
+	//
+	// This can be used in a single message to send the data, verify it and
+	// commit it.
+	COMMIT = 1 [(gogoproto.enumvalue_customname)="WriteActionCommit"];
+
+	// WriteActionAbort will release any resources associated with the write
+	// and free up the ref for a completely new set of writes.
+	ABORT = -1 [(gogoproto.enumvalue_customname)="WriteActionAbort"];
+}
+
+// WriteRequest writes data to the request ref at offset. 
+message WriteRequest {
+	// Action sets the behavior of the write.
+	//
+	// When this is a write and the ref is not yet allocated, the ref will be
+	// allocated and the data will be written at offset.
+	//
+	// If the action is write and the ref is allocated, it will accept data to
+	// an offset that has not yet been written.
+	//
+	// If the action is write and there is no data, the current write status
+	// will be returned. This works differently from status because the stream
+	// holds a lock.
+	WriteAction action = 1;
+	string ref = 2;
+	int64 expected_size = 3;
+	string expected_digest = 4 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+	int64 offset = 5;
+	bytes data = 6;
```

Comment:
First off, returning a file path for reading has a number of security issues and won't work in cases where there is no shared filesystem. Second, the suggestion about writes will create massive delays on re-hashing data contents (check out https://github.com/stevvooe/resumable for information on resuming hashes). Effectively, we can amortize the commit digesting process into each write, keeping the total write throughput at the maximum hashing bandwidth, rather than having slow commits. For a well contained content store, this coordinated writer requirement is a necessity.

In general, the approach of exposing paths is an interesting one. I very much like the prospects of doing this, but I think we need to handle the simplified cases before going forward. We can always extend this if this becomes a bottleneck. Note that this provides an API than can scale in different manners. For example, we could route requests to a content store to a cluster-enabled content cache. The assumption of a filesystem is limiting.

The main danger with doing IO in the containerd process is actually buffer bloat, which may make it more likely to spike memory usage. However, this interface is oriented to making the actual usage bounded in practice.